sshd-core/src/main/java/org/apache/sshd/client/kex/DHGClient.java

```diff
@@ -154,4 +168,54 @@ public boolean next(int cmd, Buffer buffer) throws Exception {
 
         return true;
     }
+
+    protected void verifyCertificate(Session session, OpenSshCertificate openSshKey) throws Exception {
+        PublicKey signatureKey = openSshKey.getCaPubKey();
+        String keyAlg = KeyUtils.getKeyType(signatureKey);
+        Signature verif = ValidateUtils.checkNotNull(
+                NamedFactory.create(session.getSignatureFactories(), keyAlg),
+                "No verifier located for algorithm=%s", keyAlg);
+        verif.initVerifier(session, signatureKey);
+        verif.update(session, openSshKey.getMessage());
+        if (!verif.verify(session, openSshKey.getSignature())) {
+            throw new SshException(SshConstants.SSH2_DISCONNECT_KEY_EXCHANGE_FAILED,
+                    "KeyExchange CA signature verification failed for key type=" + keyAlg);
+        }
+
+        if (openSshKey.getType() != OpenSshCertificate.SSH_CERT_TYPE_HOST) {
+            throw new SshException(SshConstants.SSH2_DISCONNECT_KEY_EXCHANGE_FAILED,
+                    "KeyExchange signature verification failed, not a host key (2): "
+                            + openSshKey.getType());
+        }
+
+        long now = System.currentTimeMillis() / 1000;
+        if (now <= openSshKey.getValidAfter() || now >= openSshKey.getValidBefore()) {
+            throw new SshException(SshConstants.SSH2_DISCONNECT_KEY_EXCHANGE_FAILED,
+                    "KeyExchange signature verification failed, CA expired: "
+                            + openSshKey.getValidAfter() + "-" + openSshKey.getValidBefore());
+        }
+
+        SocketAddress connectSocketAddress = getClientSession().getConnectAddress();
+        if (connectSocketAddress instanceof SshdSocketAddress) {
+            connectSocketAddress = ((SshdSocketAddress) connectSocketAddress).toInetSocketAddress();
+        }
+        if (connectSocketAddress instanceof InetSocketAddress) {
+            String hostName = ((InetSocketAddress) connectSocketAddress).getHostString();
+            if (!openSshKey.getPrincipals().contains(hostName)) {
```

Comment:
The code is assuming that the host name is the connect address - but what if user used an  IP address instead of a fully-qualified host name to connect ? Do we do a reverse DNS ? (I  am not too keen on it).

Reply:
The IP would have to be in the principals of the certificate in the case you want to connect with the IP. OpenSSH also does not do a reverse lookup and I also would not expect this.

Reply:
Good to know - let's place a comment in the code explaining this